File: bumperbot_cpp_examples/src/simple_tf_kinematics.cpp
#include <rclcpp/rclcpp.hpp>
#include <tf2_ros/static_transform_broadcaster.h>
#include <tf2_ros/transform_broadcaster.h>
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <chrono>

using namespace std::chrono_literals;

class SimpleTFKinematics: public rclcpp::Node
{
public:
    SimpleTFKinematics(): Node("simple_tf_kinematics")
    {
        static_transform_broadcaster = std::make_shared<tf2_ros::StaticTransformBroadcaster>(this);
        dynamic_transform_broadcaster = std::make_shared<tf2_ros::TransformBroadcaster>(this);

        geometry_msgs::msg::TransformStamped static_transform_stamped;


        static_transform_stamped.header.stamp = this->get_clock()->now();
        static_transform_stamped.header.frame_id = "base";
        static_transform_stamped.child_frame_id = "top";
        static_transform_stamped.transform.translation.x = 0.0;
        static_transform_stamped.transform.translation.y = 0.0;
        static_transform_stamped.transform.translation.z = 1.0;
        tf2::Quaternion q;
        q.setRPY(0, 0, M_PI/2);
        static_transform_stamped.transform.rotation.x = q.x();
        static_transform_stamped.transform.rotation.y = q.y();
        static_transform_stamped.transform.rotation.z = q.z();
        static_transform_stamped.transform.rotation.w = q.w();
        
        static_transform_broadcaster->sendTransform(static_transform_stamped);

        timer_ = create_wall_timer(0.1s, std::bind(&SimpleTFKinematics::timerCallback, this));


        
    }

private:
    std::shared_ptr<tf2_ros::StaticTransformBroadcaster> static_transform_broadcaster;
    std::shared_ptr<tf2_ros::TransformBroadcaster> dynamic_transform_broadcaster;

    rclcpp::TimerBase::SharedPtr timer_;
    geometry_msgs::msg::TransformStamped dynamic_transform_stamped;

    float last_x_ = 0.0;
    float x_increment_ = 0.01;

    void timerCallback()
    {
        dynamic_transform_stamped.header.stamp = this->get_clock()->now();
        dynamic_transform_stamped.header.frame_id = "odom";
        dynamic_transform_stamped.child_frame_id = "base";
        dynamic_transform_stamped.transform.translation.x = last_x_ + x_increment_;
        dynamic_transform_stamped.transform.translation.y = 0.0;
        dynamic_transform_stamped.transform.translation.z = 0.0;

        dynamic_transform_stamped.transform.rotation.x = 0.0;
        dynamic_transform_stamped.transform.rotation.y = 0.0;
        dynamic_transform_stamped.transform.rotation.z = 0.0;
        dynamic_transform_stamped.transform.rotation.w = 1.0;

        last_x_ = dynamic_transform_stamped.transform.translation.x;
        dynamic_transform_broadcaster->sendTransform(dynamic_transform_stamped);
        
        


    }
};



int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<SimpleTFKinematics>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_python_examples/bumperbot_python_examples/simple_tf_kinematics.py
import rclpy
from rclpy.node import Node
from tf2_ros.static_transform_broadcaster import StaticTransformBroadcaster
from tf2_ros import TransformBroadcaster, TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from geometry_msgs.msg import TransformStamped
from bumperbot_msgs.srv import GetTransform
from tf_transformations import quaternion_from_euler, quaternion_multiply, quaternion_inverse
import math


class SimpleTFKinematics(Node):
    def __init__(self):
        super().__init__("simple_tf_kinematics")

        self.static_tf_broadcaster = StaticTransformBroadcaster(self)
        self.dynamic_tf_broadcaster = TransformBroadcaster(self)

        self.static_transform_stamped = TransformStamped()
        self.dynamic_transform_stamped = TransformStamped()

        self.tf_buffer_ = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer_, self)

        self.x_increment_ = 0.01
        self.last_x_= 0.0
        self.last_orientation_ = quaternion_from_euler(0.0, 0.0, 0.0)
        self.orientation_increment_ = quaternion_from_euler(0.0, 0.0, 0.05)
        self.orientation_count_ = 0

        self.static_transform_stamped.header.stamp = self.get_clock().now().to_msg()
        self.static_transform_stamped.header.frame_id = "base"
        self.static_transform_stamped.child_frame_id = "top"

        self.static_transform_stamped.transform.translation.x = 0.0
        self.static_transform_stamped.transform.translation.y = 0.0
        self.static_transform_stamped.transform.translation.z = 1.0
        self.static_transform_stamped.transform.rotation.x = 0.0
        self.static_transform_stamped.transform.rotation.y = 0.0
        self.static_transform_stamped.transform.rotation.z = 0.0
        self.static_transform_stamped.transform.rotation.w = 1.0

        self.static_tf_broadcaster.sendTransform(self.static_transform_stamped)

        self.get_logger().info(f"Publishing static transform between {self.static_transform_stamped.header.frame_id}"\
                               f"and {self.static_transform_stamped.child_frame_id}")
        
        self.timer_ = self.create_timer(0.1, self.timerCallback)

        self.server_ = self.create_service(GetTransform, "get_transform", self.transformCallback)


    def timerCallback(self):    
        self.dynamic_transform_stamped.header.stamp = self.get_clock().now().to_msg()
        self.dynamic_transform_stamped.header.frame_id = "odom"
        self.dynamic_transform_stamped.child_frame_id = "base"
        
        self.dynamic_transform_stamped.transform.translation.x = self.x_increment_ + self.last_x_
        self.dynamic_transform_stamped.transform.translation.y = 0.0
        self.dynamic_transform_stamped.transform.translation.z = 0.0

        q = quaternion_multiply(self.last_orientation_, self.orientation_increment_)

        self.dynamic_transform_stamped.transform.rotation.x = q[0]
        self.dynamic_transform_stamped.transform.rotation.y = q[1]
        self.dynamic_transform_stamped.transform.rotation.z = q[2]
        self.dynamic_transform_stamped.transform.rotation.w = q[3]

        
        self.dynamic_tf_broadcaster.sendTransform(self.dynamic_transform_stamped)

        self.last_x_ = self.dynamic_transform_stamped.transform.translation.x
        self.last_orientation_ = q
        self.orientation_count_ += 1

        if self.orientation_count_ >= 100:
            self.orientation_increment_ = quaternion_inverse(self.orientation_increment_)
            self.orientation_count_ = 0

    def transformCallback(self, req, res):
        self.get_logger().info(f"Requested transform between {req.frame_id} and {req.child_frame_id}")
        requested_transform = TransformStamped()
        try:
            requested_transform = self.tf_buffer_.lookup_transform(req.frame_id, req.child_frame_id, rclpy.time.Time())
        except TransformException as e :
            res.success = False
            return res
        
        res.success = True
        res.transform = requested_transform
        return res


def main():
    rclpy.init()
    node = SimpleTFKinematics()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()